src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php

```diff
@@ -0,0 +1,443 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Controller;
+
+use Doctrine\Common\Persistence\ManagerRegistry;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+use Symfony\Component\HttpFoundation\Session\Session;
+use Symfony\Component\HttpFoundation\StreamedResponse;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Symfony\Component\HttpKernel\HttpKernelInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use Symfony\Component\Security\Csrf\CsrfToken;
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * Common features needed in controllers.
+ *
+ * The recommended way of injecting dependencies is trough getter injection.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @experimental in version 3.3
+ */
+trait ControllerTrait
+{
+    /**
+     * @required
+     */
+    protected function getRouter(): RouterInterface
+    {
+        throw new \LogicException(sprintf('An instance of "%s" must be provided.', RouterInterface::class));
```

Comment:
Done.

I've also removed the conditional behavior of the `json()` method to make the Serializer mandatory:
* it's a cleaner design (the behavior is predictable)
* it allows to not `catch` a `TypeError`
* having a proxy method doing `return new JsonResponse($data, $status, $headers)` is useless